In [1]:
import pandas as pd
import string

# Sample dataset
data = {
    'Text': [
        "I love Python!",
        "Data-Science is amazing.",
        "What's up?",
        "Clean THIS text, please!!!"
    ]
}

# Create DataFrame
df = pd.DataFrame(data)

# Function to clean text: lowercase + remove punctuation
def clean_text(text):
    text = text.lower()  # convert to lowercase
    for char in string.punctuation:
        text = text.replace(char, '')  # remove punctuation
    return text

# Apply cleaning
df['clean_text'] = df['Text'].apply(clean_text)

# Show results
print(df)


                         Text              clean_text
0              I love Python!           i love python
1    Data-Science is amazing.  datascience is amazing
2                  What's up?                whats up
3  Clean THIS text, please!!!  clean this text please


In [2]:
sample_text ="Clean THIS text, please!!!"
text = sample_text.lower()
text


'clean this text, please!!!'

In [3]:
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [4]:
for char in string.punctuation:
     text = text.replace(char, '')
text
    

'clean this text please'

In [ ]:
#library that contains punctuation
import string
string.punctuation

In [ ]:
import string

text = "Hello NLP! It's exciting, isn't it?"

result = ""  # empty string to store cleaned text

for char in text:
    if char not in string.punctuation:
        result += char  # add the character if it's NOT punctuation

print( result)


In [ ]:
result.lower()

In [ ]:
import spacy

# Load the English language model
nlp = spacy.load("en_core_web_sm")

text = "This is an example sentence to demonstrate stopword removal."

# Process the text
doc = nlp(text)

# Remove stopwords
words = []
for token in doc:
    if not token.is_stop:
        words.append(token.text)

print(words)


In [1]:
from transformers import BertTokenizer

tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
tokens = tokenizer("I loooove NLP!!! 😍", return_tensors="pt")

/home/bipin/Documents/NLP-TUTORIAL/.venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [3]:
print(tokens)

{'input_ids': tensor([[  101,  1045,  8840,  9541, 21818, 17953,  2361,   999,   999,   999,
           100,   102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}


* Example Mapping:
* Token	ID
* [CLS]	101
* I	1045
* looo (subword)	8840
* ##ove	9541
* ##e	21818
* NLP	17953
* !	2361
* !	999
* !	999
* !	999
* 😍 (unknown)	100
* [SEP]	102

* Tokenizer = automatic basic cleaner,
* but Preprocessing = intentional smart cleaner 🎯

* 🔍 Here's Why Preprocessing Is Still Important:
* 🔧 Preprocessing Task	🧠 Why It’s Useful Beyond the Tokenizer
* Remove URLs / emails	Not useful for emotion/topic models — adds noise
* Fix typos / slang	loooove may become junk tokens like 8840, 9541, etc.
* Emoji to text	😍 → [UNK]; better to convert to "happy" for emotion models
* Remove stopwords	Focus on keywords for classic ML or simple interpretable models
* Stemming / Lemmatization	Groups loved, loving, love together, useful in BOW models
* Lowercasing (for cased models)	Makes matching consistent, esp. in noisy text
* Custom filtering	Remove non-ASCII, rare tokens, domain-specific junk

* Why Preprocessing is More Important in ML (than in Transformers):
* Aspect	Classic ML Models	Transformer Models (like BERT)
* Input Type	Requires clean tokens or features	Accepts raw(ish) text → uses tokenizer
* Vocabulary Sensitivity	Very sensitive (no subword handling)	Robust (handles unknowns via subwords)
* Tokenization	Must be done manually	Handled by tokenizer
* Punctuation, Case	Can confuse the model	Often normalized internally
* Stopwords	Must be removed (they add noise)	Can be kept — model learns their usage
* Stemming/Lemmatization	Needed to reduce word variation	Often unnecessary — model understands context
* Emoji/URL Handling	Must be removed or replaced	Might still need to be cleaned (task-based)
